Welcome Dialog Service

bhavyaus opened the pull request:
Related: https://github.com/microsoft/vscode/issues/166895
The `welcomeDialogService` uses the existing https://github.com/microsoft/vscode/blob/040a68067f25066d55e98a0c87300a2837fee13d/src/vs/base/browser/ui/dialog/dialog.ts#L69

to create a welcome modal dialog. 

<img width="701" alt="image" src="https://user-images.githubusercontent.com/25044782/224198142-dbe9ed3f-0812-4716-8fc4-1cf8da0ebbb6.png">

bpasero reviewed (changes requested):
I am a little bit confused why we now have a `modalDialogService`. To clarify, we already have `IDialogService` and it opens modal dialogs (either native or custom). Why do we need a second service?

If we can move the functionality into the `IDialogService` that would be better, otherwise for a developer it will be hard to tell apart dialog service from modal dialog service.

bhavyaus commented:
> I am a little bit confused why we now have a `modalDialogService`. To clarify, we already have `IDialogService` and it opens modal dialogs (either native or custom). Why do we need a second service?
> 
> If we can move the functionality into the `IDialogService` that would be better, otherwise for a developer it will be hard to tell apart dialog service from modal dialog service.

Fixed. This dialog is specific to the welcome view. Moved and renamed service to include the `welcome` prefix.

bpasero commented on src/vs/workbench/browser/web.api.ts:
```diff
@@ -578,6 +583,34 @@ export interface IInitialColorTheme {
+	/**
+	 *  Route id.
+	 */
+	routeId: string;
```
This needs some kind of explanation

bpasero reviewed (changes requested):
Why not just `welcomeDialog`? To me, every dialog is modal by definition.

joyceerhl commented on src/vs/workbench/browser/web.api.ts:
```diff
@@ -578,6 +583,34 @@ export interface IInitialColorTheme {
+	/**
+	 *  Unique identifier of the vscode.dev route constructing the welcome dialog.
+	 */
+	routeId: string;
```
Thinking about this more, perhaps it should be renamed to `id` to be more generic across consumers, since the web api is consumed by others besides vscode.dev

bpasero replied:
But what is this used for actually?

bhavyaus replied:
I'm currently using this to check if a dialog from this route has been previously shown. I foresee its use evolving with the welcome dialog.

joyceerhl commented on src/vs/workbench/contrib/welcomeDialog/browser/welcomeDialog.contribution.ts:
```diff
@@ -0,0 +1,44 @@
+			messages: welcomeDialog.messages,
+			action: welcomeDialog.action,
+			onClose: () => {
+				storageService.store(WelcomeDialogContribution.WELCOME_DIALOG_DISMISSED_KEY + '#' + welcomeDialog.id, true, StorageScope.PROFILE, StorageTarget.MACHINE);
```
Maybe this should be `StorageTarget.USER` instead of `StorageTarget.MACHINE` so that users aren't repeatedly shown the welcome dialog if they have settings sync enabled?
